### PR TITLE
Add parser for user updates

### PR DESCRIPTION
**Pull request by sokratech**
I basically took these changes from the unity parsers and applied them to the flash parsers as I needed them

### CODE REVIEW

**Comment by sirjonasxx** [comment]
im a little confused as for why the callbacks are done in a new thread

File: g_python/htools.py [hunk 1]
```diff
@@ -33,34 +34,90 @@ def __init__(self, ext: Extension, room_users='Users', room_model='RoomReady', r
...
+                    self.__callback_remove_user(self.room_users[index])
+                del self.room_users[index]
+        finally:
+            self.__lock.release()
 
     def __load_room_users(self, message: HMessage):
         users = HEntity.parse(message.packet)
-        for user in users:
-            self.room_users[user.index] = user
+        self.__start_user_processing_thread(users)
```
Comment (sirjonasxx) [inline g_python/htools.py]: this being a thread will result in the callback happening before the updates are applied

File: g_python/htools.py [hunk 2]
```diff
@@ -33,34 +34,90 @@ def __init__(self, ext: Extension, room_users='Users', room_model='RoomReady', r
 
         self.room_users = {}
         self.__callback_new_users = None
+        self.__callback_remove_user = None
 
         self.__ext = ext
         self.__request_id = request
+        self.__lock = threading.Lock()
+        self.lock = self.__lock
```
Comment (sirjonasxx) [inline g_python/htools.py]: are you accessing this lock from the outside?

File: g_python/htools.py (hunk 1 above)
Comment (sokratech) [reply g_python/htools.py]: I basically just copied it from the Unity parsers, and didn't notice that! I will try to fix it later today or tomorrow

File: g_python/htools.py (hunk 2 above)
Comment (sokratech) [reply g_python/htools.py]: yes, I don't  know if this is a good idea or not but I needed it because there was some data race happening sometimes

**Comment by sirjonasxx** [comment]
ok i see, I'm not sure how i looked over that in the unity parser PR, but there should be no threading in the parsers/tools at all

**Comment by sokratech** [comment]
Firstly, it should be said that without multithreading, the game sometimes hangs a bit while functions are still running. This is not particularly bad, but it is noticeable. In addition, if you process the packets asynchronously, that threading can be kind of useful here. Don't nail me down on these statements though, am not very experienced and these are just my guesses based on my observations.

**Comment by sirjonasxx** [comment]
> Firstly, it should be said that without multithreading, the game sometimes hangs a bit while functions are still running. This is not particularly bad, but it is noticeable. In addition, if you process the packets asynchronously, that threading can be kind of useful here. Don't nail me down on these statements though, am not very experienced and these are just my guesses based on my observations.

Creating a thread is way more resource heavy than  just parsing it and calling listeners, so no, that is incorrect.

However, if the user who uses gpython decide to do something resource heavy, then yes. But then its not the job of the framework to create a thread for them, they should do it themself. Embedding the threading is not a good design pattern

**Comment by sokratech** [comment]
> 
> 
> > Firstly, it should be said that without multithreading, the game sometimes hangs a bit while functions are still running. This is not particularly bad, but it is noticeable. In addition, if you process the packets asynchronously, that threading can be kind of useful here. Don't nail me down on these statements though, am not very experienced and these are just my guesses based on my observations.
> 
> Creating a thread is way more resource heavy than just parsing it and calling listeners, so no, that is incorrect.
> 
> However, if the user who uses gpython decide to do something resource heavy, then yes. But then its not the job of the framework to create a thread for them, they should do it themself. Embedding the threading is not a good design pattern

That's right, as I said I have less experience myself and am happy to have learned something :D I have now removed the multithreading. Only to remove callback I still have a question: How would you solve this? Because after the del statement the object is deleted and so the HEntity object should not be available in the callback anymore. Unfortunately I have no idea how to solve it.

**Comment by sirjonasxx** [comment]
> > > Firstly, it should be said that without multithreading, the game sometimes hangs a bit while functions are still running. This is not particularly bad, but it is noticeable. In addition, if you process the packets asynchronously, that threading can be kind of useful here. Don't nail me down on these statements though, am not very experienced and these are just my guesses based on my observations.
> > 
> > 
> > Creating a thread is way more resource heavy than just parsing it and calling listeners, so no, that is incorrect.
> > However, if the user who uses gpython decide to do something resource heavy, then yes. But then its not the job of the framework to create a thread for them, they should do it themself. Embedding the threading is not a good design pattern
> 
> That's right, as I said I have less experience myself and am happy to have learned something :D I have now removed the multithreading. Only to remove callback I still have a question: How would you solve this? Because after the del statement the object is deleted and so the HEntity object should not be available in the callback anymore. Unfortunately I have no idea how to solve it.

the object is not deleted, only its reference in the dictionary is deleted. We can store the object in a local variable, then delete it from the dictionary, and then do the callback

**Comment by sokratech** [comment]
Should be fixed now

**Pull request merged**